networkx/generators/geometric.py

```diff
@@ -36,31 +29,29 @@ def euclidean(x, y):
     return sqrt(sum((a - b) ** 2 for a, b in zip(x, y)))
 
 
-def _fast_edges(G, radius, p):
-    """Returns edge list of node pairs within `radius` of each other
-       using scipy KDTree and Minkowski distance metric `p`
-
-    Requires scipy to be installed.
-    """
-    pos = nx.get_node_attributes(G, "pos")
-    nodes, coords = list(zip(*pos.items()))
-    kdtree = KDTree(coords)  # Cannot provide generator.
-    edge_indexes = kdtree.query_pairs(radius, p)
-    edges = ((nodes[u], nodes[v]) for u, v in edge_indexes)
-    return edges
-
-
-def _slow_edges(G, radius, p):
+def geometric_edges(G, radius, p):
     """Returns edge list of node pairs within `radius` of each other
-       using Minkowski distance metric `p`
 
-    Works without scipy, but in `O(n^2)` time.
+    Radius uses Minkowski distance metric `p`.
+    If scipy available, use scipy KDTree to speed computation.
     """
-    # TODO This can be parallelized.
-    edges = []
-    for (u, pu), (v, pv) in combinations(G.nodes(data="pos"), 2):
-        if sum(abs(a - b) ** p for a, b in zip(pu, pv)) <= radius ** p:
-            edges.append((u, v))
+    try:
+        from scipy.spatial import cKDTree as KDTree
+    except ImportError:
+        KDTree = None
+    nodes_pos = G.nodes(data="pos")
+    if KDTree is not None:
+        nodes, coords = list(zip(*nodes_pos))
+        kdtree = KDTree(coords)  # Cannot provide generator.
+        edge_indexes = kdtree.query_pairs(radius, p)
+        edges = ((nodes[u], nodes[v]) for u, v in edge_indexes)
+        return edges
+    # no scipy KDTree so compute by for-loop
```

Comment:
I have a few comments here: 

First, small issue, but I would have put this in an else block. I like to avoid early returns when possible (makes it easier to copy-paste in IPython) and it also makes it clear that this is an alternative to the above code. 

Second, computing `radius ** p` will be executed on every iteration of this loop. It would be more efficient to do this once `p_thresh = radius ** p` and then using `p_thresh` instead. 

Demo showing how python does not precompute this value:

```python
class MyValue:
    def __init__(self, value):
        self.value = value

    def __pow__(self, arg):
        print('EXECUTE OP')
        return self.__class__(self.value ** arg)


def main():
    x = MyValue(3)
    [i for i in range(10) if x ** 3]

if __name__ == '__main__':
    main()
```

The `print('EXECUTE OP')` executes 10 times.

Lastly, this definition of `edges` is a list whereas the one in the KD-Tree version is a generator. It should likely be changed to be consistent.

Reply:
These are good suggestions..  Thanks!